### ISSUE TITLE
Error running mentalist on a recently build DB

### ISSUE THREAD

**Issue by crarlus** [WGS-TB/MentaLiST#71]
I am running mentalist with the most recent conda version (0.2.2)
I successfully build an DB (based on Enterobase, manually Downloaded by me). However when I try to run `mentalist call`, I get the following error:

```
$ mentalist call -o user/testdata/QCumber_run/mentalist/11-00224_newDB_v3.mlst --db /cephfs/abteilung4/NGS/ReferenceDB/MLST/Enterobase/senterica/cgMLSTv2/20180625/senterica_enterobase_cgmlst_31.mentalist.db -1 11-00224_S15_L001.1P.fastq.gz -2 11-00224_S15_L001.2P.fastq.gz 2>&1 | tee user/testdata/QCumber_run/mentalist/11-00224_newDB_v3.log

2018-06-25T12:40:32.623 - info: Opening kmer database ...
2018-06-25T12:42:35.372 - info: Finished the JLD load.
2018-06-25T12:44:46.417 - info: Building kmer index ...
2018-06-25T12:58:59.702 - info: Sample: 11-00224_S15_L001. Opening fastq file(s) and counting kmers ...
2018-06-25T13:00:35.358 - info: Voting for alleles ...
2018-06-25T13:01:52.664 - info: Calling alleles and novel alleles ...
ERROR: LoadError: GZip.GZError(-1, "gzopen failed")
Stacktrace:
 [1] gzopen(::String, ::String, ::Int64) at user/anaconda3/envs/mentalist/share/julia/site/v0.6/GZip/src/GZip.jl:250
 [2] FastaIO.FastaReader{String}(::String) at user/anaconda3/envs/mentalist/share/julia/site/v0.6/FastaIO/src/FastaIO.jl:48
 [3] read_alleles(::String, ::Set{Int16}) at user/anaconda3/envs/mentalist/bin/calling_functions.jl:175
 [4] (::#call_allele#31{Int64,DataStructures.DefaultDict{Bio.Seq.Kmer{Bio.Seq.DNANucleotide,31},Int64,Int64},Int64,Int64})(::Dict{Int16,Int64}, ::Int64, ::String, ::Array{Int16,1}, ::String) at user/anaconda3/envs/mentalist/bin/calling_functions.jl:221
 [5] (::##30#50{Dict{Int16,Dict{Int16,Int64}},DataStructures.DefaultDict{Int16,Int64,Int64},Array{String,1},Dict{Int16,Array{Int16,1}},Array{Any,1},#call_allele#31{Int64,DataStructures.DefaultDict{Bio.Seq.Kmer{Bio.Seq.DNANucleotide,31},Int64,Int64},Int64,Int64}})(::Tuple{Int64,String}) at ./<missing>:0
 [6] collect_to!(::Array{AlleleCall,1}, ::Base.Generator{Enumerate{Array{String,1}},##30#50{Dict{Int16,Dict{Int16,Int64}},DataStructures.DefaultDict{Int16,Int64,Int64},Array{String,1},Dict{Int16,Array{Int16,1}},Array{Any,1},#call_allele#31{Int64,DataStructures.DefaultDict{Bio.Seq.Kmer{Bio.Seq.DNANucleotide,31},Int64,Int64},Int64,Int64}}}, ::Int64, ::Tuple{Int64,Int64}) at ./array.jl:508
 [7] collect(::Base.Generator{Enumerate{Array{String,1}},##30#50{Dict{Int16,Dict{Int16,Int64}},DataStructures.DefaultDict{Int16,Int64,Int64},Array{String,1},Dict{Int16,Array{Int16,1}},Array{Any,1},#call_allele#31{Int64,DataStructures.DefaultDict{Bio.Seq.Kmer{Bio.Seq.DNANucleotide,31},Int64,Int64},Int64,Int64}}}) at ./array.jl:476
 [8] call_alleles(::Int64, ::DataStructures.DefaultDict{Bio.Seq.Kmer{Bio.Seq.DNANucleotide,31},Int64,Int64}, ::Dict{Int16,Dict{Int16,Int64}}, ::DataStructures.DefaultDict{Int16,Int64,Int64}, ::Array{String,1}, ::Dict{Int16,Array{Int16,1}}, ::Array{Any,1}, ::Int64, ::Int64, ::Bool) at user/anaconda3/envs/mentalist/bin/calling_functions.jl:327
 [9] run_calling_pipeline(::Dict{String,Any}) at user/anaconda3/envs/mentalist/bin/calling_functions.jl:32
 [10] call_mlst(::Dict{String,Any}) at user/anaconda3/envs/mentalist/bin/mentalist:185
while loading user/anaconda3/envs/mentalist/bin/mentalist, in expression starting on line 247
```

**Comment by dfornika** [comment]
Thanks for this report. Could you run the `strings` command on the database and  pipe it to `grep fasta_files` to see the fasta paths that are stored in the database? There will be many entries, could you post just a few of them here?

**Comment by crarlus** [comment]
Hi, thanks for your support. Here is what I get:
```
$ strings senterica_enterobase_cgmlst_31.mentalist.db | grep fasta_files |  head

{"profile":"user/Enterobase/senterica/cgMLSTv2/20180625/cgMLST_v2/MLST-profiles","threads":2,"k":31,"fasta_files":["alleles_cgMLST_v2/SC0831.fasta","alleles_cgMLST_v2/SEN0401.fasta","alleles_cgMLST_v2/SNSL254_A1382.fasta","alleles_cgMLST_v2/SPAB_04503.fasta
```
As far as I can see, it has the same structure as the previous built:
````
$ strings senterica_new.db | grep fasta_files | tr ',' '\n' | head
{"profile":"SALwgMLST.cgMLSTv1.profiles.list"
"threads":2
"k":31
"fasta_files":["fasta/SC0831.fasta"
"fasta/SEN0401.fasta"
"fasta/SNSL254_A1382.fasta"
"fasta/SPAB_04503.fasta"
"fasta/STM0834.fasta"
"fasta/STM0942.fasta"
"fasta/STM1092.fasta"
```

**Comment by dfornika** [comment]
It does seem like it's an issue with reading the scheme fasta files.

The error seems to be triggered here: https://github.com/WGS-TB/MentaLiST/blob/bbb2381b06b1509b5645bb09bc027f27ad6a99a2/src/calling_functions.jl#L175

Can you confirm that the paths that are stored in the `fasta_files` section of your database are the correct relative paths to the fasta files? For example, for your `senterica_new.db` file, it should be stored alongside a directory called `fasta` that contains the scheme fasta files.

For your `senterica_enterobase_cgmlst_31.mentalist.db` database, it should be stored alongside a directory called `alleles_cgMLST_v2` that contains the scheme fasta files.

**Comment by dfornika** [comment]
Thanks for your detailed answer. I'll try to reproduce the error on my system so we can dig into the problem more closely. I'm out at a conference for the next few days but I'll try to find some time to work on this.

@pedrofeijao do you have some time to look at this as well? Can you see what the problem is?

**Comment by pedrofeijao** [comment]
I made some tests and could not reproduce the error. It is probably not
related to the file locations, since mentalist checks if the fasta files
exist and exits with a proper message when they don't. The fast that is a
gzip error is also strange, since no gzip function is called directly at
the line of the error.

**Comment by dfornika** [comment]
I'm having trouble copy/pasting from my terminal so here is a screenshot of what the output looked like:

<img width="1427" alt="screen shot 2018-06-30 at 9 23 44 am" src="https://user-images.githubusercontent.com/145659/42127104-b2223b74-7c47-11e8-89ba-1f9eb48533c5.png">

If I run `strings s_enterica_enterobase_partial_k31.db | grep fasta_files` then I see this:
```
{"profile":null,"threads":4,"k":31,"fasta_files":["s_enterica_enterobase_partial/SC0831.fasta","s_enterica_enterobase_partial/SEN0401.fasta","s_enterica_enterobase_partial/SNSL254_A1382.fasta","s_enterica_enterobase_partial/SPAB_04503.fasta"],"db":"s_enterica_enterobase_partial_k31.db"}
```

**Comment by dfornika** [comment]
One further thing that I tried: I renamed one of the scheme fasta files so that it no longer matched the path stored in the kmer database. Then I ran `mentalist call` again. I got this output:

```
(mentalist-0.2.2) [dfornika@sabin mentalist-testing]$ mentalist call --db databases-0.2.2/s_enterica/s_enterica_enterobase_partial_k31.db -o output-0.2.2/SIM_CFSAN000189 -1 input/SIM_CFSAN000189_1.fastq.gz -2 input/SIM_CFSAN000189_2.fastq.gz 
WARNING: imported binding for VERSION overwritten in module Main
WARNING: Method definition (::Type{Bio.Align.SAMWriter{T} where T<:IO})(IO) in module Align at /home/dfornika/miniconda3/envs/mentalist-0.2.2/share/julia/site/v0.6/Bio/src/align/hts/sam/writer.jl:14 overwritten at /home/dfornika/miniconda3/envs/mentalist-0.2.2/share/julia/site/v0.6/Bio/src/align/hts/sam/writer.jl:22.

WARNING: deprecated syntax "abstract SeqRead" at /home/dfornika/miniconda3/envs/mentalist-0.2.2/share/julia/site/v0.6/OpenGene/src/Base/Types/read.jl:2.
Use "abstract type SeqRead end" instead.

WARNING: deprecated syntax "typealias GtfHeader OrderedDict{ASCIIString,ASCIIString}" at /home/dfornika/miniconda3/envs/mentalist-0.2.2/share/julia/site/v0.6/OpenGene/src/Base/Types/gtf.jl:5.
Use "const GtfHeader = OrderedDict{ASCIIString,ASCIIString}" instead.

WARNING: deprecated syntax "typealias GtfData Array{GtfItem,1}" at /home/dfornika/miniconda3/envs/mentalist-0.2.2/share/julia/site/v0.6/OpenGene/src/Base/Types/gtf.jl:32.
Use "const GtfData = Array{GtfItem,1}" instead.
2018-06-30T09:42:32.543 - info: Opening kmer database ... 
2018-06-30T09:42:34.555 - info: Finished the JLD load.
2018-06-30T09:42:34.825 - info: Building kmer index ...
2018-06-30T09:42:35.068 - error: The following input file(s) could not be found: databases-0.2.2/s_enterica/s_enterica_enterobase_partial/SC0831.fasta.
exiting ...
```

**Comment by crarlus** [comment]
Hi,
some updates
* It does work with a number of other data sources (cgmlst.org and pubmlst.org)
* I added the suggested print option and replaced the last print fasta file before the error with an older version. Rebuilding the entire DB and calling mentalist again throws an error at the next fasta file. So apparently, there is something odd with the files
* I did get it to work on a different computer. I will try to evaluate the differences between the files and the system config

**Comment by pedrofeijao** [comment]
Hi,
an update on this issue; I was testing MentaLiST on many large cg and wgMLST schemes and was sometimes getting this error message. Apparently this is caused by FastaIO calling the gzip function, which is odd since at this point we're reading regular fasta files, not gzipped. I changed the fasta reading library to a different one, OpenGene, which was already being used to read FASTQ files, and I'm not getting the error again, so I think this is most likely fixed and I am closing the issue for now. 
Closed by commit 24ce14532f21084ed601d60345d1bdc2eeb5a5e0
